lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+       a request to acquire the lock, and a request to release the lock, respectively.
+
+    2) The application data in the relation. This tracks whether the lock has been
+       "granted", Or has been released (and reverted to idle). There are two valid states:
+       "granted" or None.  If a lock is in the "granted" state, a unit should emit a
+       RunWithLocks event and then release the lock.
+
+       If a lock is in "None", this means that a unit has not yet requested the lock, or
+       that the request has been completed.
+
+    In more detail, here is the relation structure:
+
+    relation.data:
+        <unit n>:
+            status: 'acquire|release'
+        <application>:
+           <unit n>: 'granted|None'
+
+    Note that this class makes no attempts to timestamp the locks and thus handle multiple
+    requests in a row. If a unit re-requests a lock before being granted the lock, the
+    lock will simply stay in the "acquire" state. If a unit wishes to clear its lock, it
+    simply needs to call lock.release().
+
+    """
+
+    # Track lock requests from units.
+    ACQUIRE = 'acquire'
+    RELEASE = 'release'
+
+    # Track responses from the leader.
+    GRANTED = 'granted'
+    IDLE = 'idle'
+
+    def __init__(self, rel_name, charm, unit=None):
+        self.relation = charm.model.relations[rel_name][0]
+        if not self.relation:
+            # TODO: defer caller in this case (probably just fired too soon).
+            raise LockErrorNoRelation()
+
+        self.unit = unit or charm.framework.model.unit
+        self.app = charm.framework.model.app
+
+    @property
+    def status(self):
+        """Return an appropriate status.
+
+        Note that the state exists in the unit's relation data, and the application
+        relation data, so we have to be careful about what our states mean.
+
+        Unit status can only be in "acquire", "release", "None" (None means unset)
+        Application status can only be in "granted" or "None" (None means unset or released)
+
+        """
+        unit_status = self.relation.data[self.unit].get('status', None)
+        app_status = self.relation.data[self.app].get(self.unit, self.IDLE)
+
+        if app_status == self.GRANTED and unit_status == self.RELEASE:
+            # Active release request.
+            return self.RELEASE
+
+        if app_status == self.IDLE and unit_status == self.ACQUIRE:
+            # Active acquire request.
+            return self.ACQUIRE
+
+        return app_status  # Granted or unset/released
+
+    @status.setter
+    def status(self, status):
+        """Set the given status.
+
+        Since we update the relation data, this may fire off a RelationChanged event.
+        """
+        if status == self.ACQUIRE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
```

Comment:
```suggestion
        if status == self.ACQUIRE or status == self.RELEASE:
            self.relation.data[self.unit].update({"status": status})
```

Reply:
+1 (I've got local changes, so I'm not going to merge just yet.)